### PR TITLE
Do not change the CSRF token cookie if the response is not successful

### LINKED ISSUE #2246: CSRF-Token kommt durcheinander und ein Login ist nicht mehr möglich
**Affected version(s)**

Contao 4.9

**Description**

In einem Projekt ist uns jetzt aufgefallen, dass der CSRF-Mechanismus durcheinander gebracht werden kann, so dass kein Login für mit Mitglieder mehr möglich ist. Aufgefallen war uns das beim IE11, das lässt sich aber auch bei anderen reproduzieren, wenn auch nicht so einfach. Ausgangspunkt:

- Eine Site mit Mitgliedern.
- eine 404-Umleitung auf ein Seite (z.b. Startseite oder Loginseite)

Wenn es jetzt auf der Seite ein fehlendes Element gibt, dann wird dafür die 404er geliefert, da da aber ein Login-Formular integriert ist wird das Token erneuert und das "richtige" in der Seite funktioniert nicht mehr.

Uns ist das aufgefallen, da die SourceMaps einiger JS-Files nicht mit deployed waren und der IE11 dafür dann die startseite.html geladen hatte. Die anderen Browser ignorieren wohl den 404er an der Stelle.

Und nein, das ist vermutlich nicht nur auf den IE beschränkt. Wie schnell kann es durch Unachtsamkeit passieren, dass ein Image im CSS fehlt. Dabei gehe ich mal davon aus, dass auch die anderen dem Link folgen würden und da das selbe passiert. Ich hatte im Chrome auch ein paar mal die CSRF-Fehlermeldung. Bin dem aber damals nicht nachgegangen, da ein reload das behoben hatte.

Volker

### PR DESCRIPTION
**Pull request by fritzmg**
| Q                | A
| -----------------| ---
| Fixed issues     | Fixes #2246
| Docs PR or issue | -

I am not sure if this is the right approach to fix #2246, but I think it makes sense to never change the CSRF token cookie, if the response is not within the `>= 200, < 300` status code range. Implementing this check would fix #2246 as far as I tested.

The problem is, that an AJAX request on the same page that requires the presence of a CSRF token cookie can invalidate said CSRF token cookie, if the generated response in turn requires a new CSRF token cookie (e.g., if the 404 page redirects to another Contao page which in turn requires a CSRF token cookie).

The problem does not occur in Contao 4.4, at least not with the reproduction I posted.

### CODE REVIEW

**Comment by ausi** [comment]
I think this might break the CSRF cookie creation for login pages with a 401 status code.

Would moving the `isSuccessful()` check down before the line `$this->removeCookies(…);` also solve the issue? So that we only remove the CSRF cookie in successful responses but still set the cookie in unsuccessful ones?

**Comment by fritzmg** [comment]
> I think this might break the CSRF cookie creation for login pages with a 401 status code.

Yeah, you are right, this is not the correct solution. It does break forms on the 401 page of course.

> Would moving the `isSuccessful()` check down before the line `$this->removeCookies(…);` also solve the issue? So that we only remove the CSRF cookie in successful responses but still set the cookie in unsuccessful ones?

That seems to work, though I don't fully understand why.

Also during testing I noticed, that `$request->isXmlHttpRequest()` is never true, even though it is an `XMLHttpRequest`.

**Comment by ausi** [comment]
> Also during testing I noticed, that `$request->isXmlHttpRequest()` is never true, even though it is an `XMLHttpRequest`.

This is only true if you send the header `X-Requested-With: XMLHttpRequest` (which most javascript frameworks do, including the one of Contao).

File: core-bundle/src/EventListener/CsrfTokenCookieSubscriber.php [hunk 1]
```diff
@@ -71,6 +71,11 @@ public function onKernelResponse(ResponseEvent $event): void
         if ($this->requiresCsrf($request, $response)) {
             $this->setCookies($request, $response);
         } else {
+            // Do not delete CSRF token cookie, if response is not successful (#2252)
+            if (!$response->isSuccessful()) {
+                return;
+            }
+
```
Comment (ausi) [inline core-bundle/src/EventListener/CsrfTokenCookieSubscriber.php]:
```suggestion
        } elseif ($response->isSuccessful()) {
            // Only delete CSRF token cookie, if response is successful (#2252)
```

**Comment by ausi** [comment]
> That seems to work, though I don't fully understand why.

Setting the request token cookie should not be a problem in most cases because it is just ignored if it is present but not needed. But deleting an existing request token cookie can lead to problems if it is missing.

**Comment by aschempp** [comment]
@ausi's change seems right to me.

**Review by aschempp — CHANGES_REQUESTED** [review]
@ausi's change suggestion seems right to me.

File: core-bundle/src/EventListener/CsrfTokenCookieSubscriber.php (hunk 1 above)
Comment (fritzmg) [reply core-bundle/src/EventListener/CsrfTokenCookieSubscriber.php]: I am not really in favor of `elseif` - here and in general. A separate condition makes it more readable imho.

Comment (leofeyer) [reply core-bundle/src/EventListener/CsrfTokenCookieSubscriber.php]: No, it does not. 🙈 Even my IDE says "this condition can be merged into the parent one". 😄

File: core-bundle/src/EventListener/CsrfTokenCookieSubscriber.php [hunk 2]
```diff
@@ -70,7 +70,8 @@ public function onKernelResponse(ResponseEvent $event): void
 
         if ($this->requiresCsrf($request, $response)) {
             $this->setCookies($request, $response);
-        } else {
+        } elseif ($response->isSuccessful()) {
+            // Only delete CSRF token cookie, if response is successful (#2252)
```
Comment (leofeyer) [inline core-bundle/src/EventListener/CsrfTokenCookieSubscriber.php]:
```suggestion
            // Only delete the CSRF token cookie if the response is successful (#2252)
```

There is no comma before "if" if the dependent clause follows the independent one. 🤓

**Review by leofeyer — CHANGES_REQUESTED** [review]

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @fritzmg.

**Pull request merged**